src/nias_cpp/interfaces/inner_products.h

```diff
@@ -0,0 +1,155 @@
+        bool pairwise = false, const std::optional<Indices>& left_indices = std::nullopt,
+        const std::optional<Indices>& right_indices = std::nullopt) const
+    {
+        const auto ret = apply(left, right, pairwise, left_indices, right_indices);
+        if (pairwise)
+        {
+            return pybind11::array(ret.size(), ret.data());
+        }
+
+        const ssize_t n = left_indices ? left_indices->size(left.size()) : left.size();
+        const ssize_t m = right_indices ? right_indices->size(right.size()) : right.size();
+        pybind11::array_t<F> ret_array({n, m});
+        auto ret_array_mutable = ret_array.mutable_unchecked();
+        for (ssize_t i = 0; i < n; ++i)
+        {
+            for (ssize_t j = 0; j < m; ++j)
+            {
+                ret_array_mutable(i, j) = ret[(i * m) + j];
+            }
+        }
+        return ret_array;
+    }
+};
+
+template <floating_point_or_complex F>
+class NormInterface
+{
+   public:
+    using ScalarType = F;
+
+    NormInterface() = default;
+    NormInterface(const NormInterface&) = default;
+    NormInterface(NormInterface&&) = default;
+    NormInterface& operator=(const NormInterface&) = default;
+    NormInterface& operator=(NormInterface&&) = default;
+    virtual ~NormInterface() = default;
+
+    /**
+     * @brief Compute the norms of the vectors in the vector array.
+     * @returns A numpy array containing the norms of the vectors in the array.
+     */
+    virtual std::vector<ScalarType> operator()(const VectorArrayInterface<ScalarType>& vec_array) const = 0;
+};
+
+// forward declaration
+template <floating_point_or_complex F>
+class InnerProductInterface;
+
+template <floating_point_or_complex F>
+class InnerProductBasedNorm : public NormInterface<F>
+{
+   public:
+    explicit InnerProductBasedNorm(const InnerProductInterface<F>& inner_product)
+        : inner_product_(inner_product)
+    {
+    }
+
+    std::vector<F> operator()(const VectorArrayInterface<F>& vec_array) const override
+    {
+        auto norms = inner_product_.apply(vec_array, vec_array, true);
+        for (ssize_t i = 0; i < vec_array.size(); ++i)
+        {
+            norms[i] = std::sqrt(norms[i]);
+        }
+        return norms;
+    }
+
+   private:
+    const InnerProductInterface<F>& inner_product_;
+};
+
+template <floating_point_or_complex F>
+class InnerProductInterface : public SesquilinearFormInterface<F>
+{
+   public:
+    InnerProductInterface()
+        : induced_norm_(std::make_unique<InnerProductBasedNorm<F>>(*this)) {};
+
+    InnerProductInterface(const InnerProductInterface&) = delete;
+    InnerProductInterface(InnerProductInterface&&) = default;
+    InnerProductInterface& operator=(const InnerProductInterface&) = delete;
+    InnerProductInterface& operator=(InnerProductInterface&&) = default;
+    virtual ~InnerProductInterface() = default;
+
+    virtual const NormInterface<F>& norm() const
```

Comment:
I'd rename that to
```suggestion
    virtual const NormInterface<F>& induced_norm() const
```
to follow https://github.com/nias-project/nias/blob/main/src/nias/interfaces.py#L277C9-L277C21